### PR TITLE
chore: re-enable HW processing for hermetic build scripts

### PR DESCRIPTION
**Pull request by diegomarquezp**
re-enables the changes from https://github.com/googleapis/sdk-platform-java/pull/2342 as part of [milestone 3.2](https://docs.google.com/document/d/1v-sJBmdNEVBRryL8n90XK8OIiqVphaJemXVhANw14Jk/edit?pli=1&resourcekey=0-QGUb2do-JBlDWKvptWBp_g&tab=t.0#bookmark=id.kl5fzqdav6u7) of Hermetic Build

### Changes:
 - add support for HW libs in `integration_tests.py`
 - adapt configuration of docker volumes
 - adapt usage of `fix-poms.py` to use monorepo and split_repo modes
 - fix temp_destination_path in `generate_composed_library.py` in order to avoid collisions
   - For example `google/bigtable/v2` and `google/bigtable/admin/v2` would both be sent to a temporal `java-bigtable-v2` folder, spoiling the files being transferred via `copy-code`. 
   - The fix is to have a `proto_path`-based folder name to achieve uniqueness

### After this PR:
 - [ ] Reconfigure the nightly CI in google-cloud-java to use the updated volume mapping

### CODE REVIEW

File: library_generation/owlbot/bin/entrypoint.sh [hunk 1]
```diff
@@ -78,16 +79,7 @@ echo "Fixing missing license headers..."
 python3 "${scripts_root}/owlbot/src/fix-license-headers.py"
 echo "...done"
 
-# TODO: re-enable this once we resolve thrashing
```
Comment (diegomarquezp) [inline library_generation/owlbot/bin/entrypoint.sh:81]: This was meant to restore the license header year (e.g. keep `2023` if the generated license now shows `2024`) by looking at the logs (owlbot previously worked in a git repo but we now don't do it anymore). This is just an additional cleanup

File: library_generation/test/resources/goldens/.repo-metadata-golden.json [hunk 2]
```diff
@@ -7,7 +7,7 @@
   "release_level": "preview",
   "transport": "grpc",
   "language": "java",
-  "repo": "googleapis/google-cloud-java",
+  "repo": "java-bare-metal-solution",
```
Comment (JoeWang1127) [inline library_generation/test/resources/goldens/.repo-metadata-golden.json:10]: Why do we need to change this golden file?

File: library_generation/model/library_config.py [hunk 3]
```diff
@@ -46,6 +46,7 @@ def __init__(
         rpc_documentation: Optional[str] = None,
         cloud_api: Optional[bool] = True,
         requires_billing: Optional[bool] = True,
+        extra_versioned_modules: Optional[str] = None,
```
Comment (JoeWang1127) [inline library_generation/model/library_config.py:49]: What is the meaning of this parameter?

File: library_generation/test/resources/goldens/.repo-metadata-golden.json (hunk 2 above)
Comment (diegomarquezp) [reply library_generation/test/resources/goldens/.repo-metadata-golden.json:10]: From https://github.com/googleapis/sdk-platform-java/blob/69c705b537ae93b18deb2ebb34bbeec91fa3df15/library_generation/test/unit_tests.py#L379-L387

The unit test creates a `GenerationConfig` with one library only (not monorepo). This makes the `repo` entry to be treated as if it was a standalone HW library

File: library_generation/model/library_config.py (hunk 3 above)
Comment (diegomarquezp) [reply library_generation/model/library_config.py:49]: It's just another `.repo-metadata` possible entry that `java-bigtable` [uses](https://github.com/googleapis/java-bigtable/blob/5f246ef47fe4e92ffd0387c4c87ed64b5e789d72/.repo-metadata.json#L15)

Comment (diegomarquezp) [reply library_generation/model/library_config.py:49]: Apparently it's unofficial [[1]](https://docs.google.com/document/d/1IUpyM_V-LOC3OUSlkcG5GgqNzjQ5qnnlkEF9gu2kJR4/edit?tab=t.0#heading=h.x9snb54sjlu9) [[2]](https://docs.google.com/document/d/1Dbq-ke1B0LqXs0AS17FNpm2U0yOrefHssCDuzFqBXJI/edit?tab=t.0#heading=h.jyjlyheps42e)

Comment (diegomarquezp) [reply library_generation/model/library_config.py:49]: > meaning

It is used to add certain modules for pom processing in `fix-poms.py`
https://github.com/googleapis/sdk-platform-java/blob/7c09e230b6528a4f8c4590fcc4ab681b7a98eac7/library_generation/owlbot/src/fix-poms.py#L320-L323

File: library_generation/test/resources/goldens/.repo-metadata-golden.json (hunk 2 above)
Comment (diegomarquezp) [reply library_generation/test/resources/goldens/.repo-metadata-golden.json:10]: The core logic is in https://github.com/googleapis/sdk-platform-java/blob/69c705b537ae93b18deb2ebb34bbeec91fa3df15/library_generation/utilities.py#L347-L351

**Comment by JoeWang1127** [comment]
Can you link the corresponding task in project plan in PR description?

File: library_generation/utilities.py [hunk 4]
```diff
@@ -407,6 +398,8 @@ def generate_prerequisite_files(
         repo_metadata["rest_documentation"] = library.rest_documentation
     if library.rpc_documentation:
         repo_metadata["rpc_documentation"] = library.rpc_documentation
+    if library.extra_versioned_modules:
+        repo_metadata["extra_versioned_modules"] = library.extra_versioned_modules
```
Comment (JoeWang1127) [inline library_generation/utilities.py:402]: Do you think we should add an unit test to verify this assignment?

File: library_generation/test/integration_tests.py [hunk 5]
```diff
@@ -73,9 +82,12 @@ def test_generate_repo(self):
                     f"with the actual library in {repo_dest}/{library_name}. "
```
Comment (JoeWang1127) [inline library_generation/test/integration_tests.py:82]: Should we change the actual library path in log?

Comment (diegomarquezp) [reply library_generation/test/integration_tests.py:82]: Indeed

File: library_generation/utilities.py (hunk 4 above)
Comment (diegomarquezp) [reply library_generation/utilities.py:402]: I added an `extra_versioned_modules` entry in `unit_tests.library_1` and updated the golden `.repo-metadata.json`s. I think that should cover it

File: library_generation/test/integration_tests.py (hunk 5 above)
Comment (diegomarquezp) [reply library_generation/test/integration_tests.py:82]: Fixed

**Comment by diegomarquezp** [comment]
> Can you link the corresponding task in project plan in PR description?

Done

File: library_generation/test/resources/goldens/.repo-metadata-golden.json (hunk 2 above)
Comment (diegomarquezp) [reply library_generation/test/resources/goldens/.repo-metadata-golden.json]: Added a second unit test

File: library_generation/utilities.py [hunk 6]
```diff
@@ -356,7 +345,9 @@ def generate_prerequisite_files(
     )
     distribution_name_short = re.split(r"[:/]", distribution_name)[-1]
     repo = (
-        "googleapis/google-cloud-java" if is_monorepo else f"{language}-{library_name}"
+        "googleapis/google-cloud-java"
+        if config.is_monorepo
+        else f"{language}-{library_name}"
```
Comment (JoeWang1127) [inline library_generation/utilities.py:350]: I think this should be `f"googleapis/{language}-{library_name}"`

**Review by JoeWang1127 — APPROVED** [review]

**Review by suztomo — APPROVED** [review]

File: library_generation/test/resources/goldens/.repo-metadata-non-monorepo-golden.json [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+  "api_description": "Bring your Oracle workloads to Google Cloud with Bare Metal Solution and jumpstart your cloud journey with minimal risk.",
+  "client_documentation": "https://cloud.google.com/java/docs/reference/google-cloud-bare-metal-solution/latest/overview",
+  "release_level": "preview",
+  "transport": "grpc",
+  "language": "java",
+  "repo": "googleapis/java-bare-metal-solution",
+  "repo_short": "java-bare-metal-solution",
+  "distribution_name": "com.google.cloud:google-cloud-bare-metal-solution",
+  "api_id": "baremetalsolution.googleapis.com",
+  "library_type": "GAPIC_AUTO",
```
Comment (JoeWang1127) [inline library_generation/test/resources/goldens/.repo-metadata-non-monorepo-golden.json:14]: Do we need to change this value to `GAPIC_COMBO`?

Comment (diegomarquezp) [reply library_generation/test/resources/goldens/.repo-metadata-non-monorepo-golden.json:14]: Just hardcode it in the test

File: library_generation/generate_repo.py [hunk 8]
```diff
@@ -102,6 +108,11 @@ def generate_from_yaml(
             versions_file=repo_config.versions_file,
         )
 
+    # we skip repo_level_postprocess if not in a monorepo since it's meant for
```
Comment (blakeli0) [inline library_generation/generate_repo.py:111]: I know it was not introduced in this PR, can we rename `repo_level_postprocess` to `monorepo_postprocess` if it's meant for monorepo only?

File: library_generation/model/library_config.py (hunk 3 above)
Comment (blakeli0) [reply library_generation/model/library_config.py:49]: @diegomarquezp Going forward, I'm not sure accommodating every non-standard field in repo-metadata.json in our scripts is the way to go, we may want to leave the responsibility to each handwritten library owner. For now, can you please document this new parameter in [go/java-hermetic-build-config-spec](http://goto.google.com/java-hermetic-build-config-spec)?

File: library_generation/test/resources/goldens/.repo-metadata-monorepo-golden.json [hunk 9]
```diff
@@ -14,5 +14,6 @@
   "library_type": "GAPIC_AUTO",
   "requires_billing": true,
   "rest_documentation": "https://cloud.google.com/bare-metal/docs/reference/rest",
-  "rpc_documentation": "https://cloud.google.com/bare-metal/docs/reference/rpc"
+  "rpc_documentation": "https://cloud.google.com/bare-metal/docs/reference/rpc",
+  "extra_versioned_modules": "test-module"
```
Comment (blakeli0) [inline library_generation/test/resources/goldens/.repo-metadata-monorepo-golden.json:18]: This parameter should not be in monorepo goldens. Do `google-cloud-java` and `java-bigtable` share the same test setup?

Comment (diegomarquezp) [reply library_generation/test/resources/goldens/.repo-metadata-monorepo-golden.json:18]: I treated this repo-metadata entry as monorepo-agnostic, but I agree it's closer to real life to not include it here. I removed it and adapted the test accordingly.

File: library_generation/generate_repo.py (hunk 8 above)
Comment (diegomarquezp) [reply library_generation/generate_repo.py]: Agreed. Renamed to `monorepo_postprocessing`

File: library_generation/test/resources/goldens/.repo-metadata-non-monorepo-golden.json (hunk 7 above)
Comment (diegomarquezp) [reply library_generation/test/resources/goldens/.repo-metadata-non-monorepo-golden.json]: Done

File: library_generation/model/library_config.py (hunk 3 above)
Comment (diegomarquezp) [reply library_generation/model/library_config.py:49]: I added it to the document. What would it mean for a HW library owner to maintain this entry? They do need that entry in their `.repo-metadata` and we are now making this json file as generated and having the new yaml as source of truth, for which we will need all these non-standard fields (that affect the output library) to be added in the yaml-to-json logic.

Comment (blakeli0) [reply library_generation/model/library_config.py:49]: For example, we don't have to generate `.repo-metadata.json` if it's a split repo, and HW library owners can continue maintain them manually as they are today.

**Review by blakeli0 — APPROVED** [review]

**Pull request merged**